core/src/test/java/com/sequenceiq/cloudbreak/service/stack/flow/UpdateNodeCountValidatorTest.java

```diff
@@ -246,6 +250,80 @@ private static Stream<Arguments> testValidateStatusForStopStartHostGroupData() {
+        instanceMetaData.setInstanceMetadataType(instanceMetadataType);
+        instanceMetaData.setInstanceStatus(instanceStatus);
+        instanceMetaData.setDiscoveryFQDN("hg0-host-2");
+        return instanceMetaData;
+    }
+
+    @Test
+    public void testValidateCMStatusNonPrimaryGatewayUnhealthy() {
+        StackDto stack = mock(StackDto.class);
+        List<InstanceMetadataView> instanceMetaDataAsList = new ArrayList<>();
+        instanceMetaDataAsList.add(generateInstanceMetadata(InstanceStatus.SERVICES_HEALTHY, InstanceMetadataType.GATEWAY_PRIMARY));
+        instanceMetaDataAsList.add(generateInstanceMetadata(InstanceStatus.SERVICES_UNHEALTHY, InstanceMetadataType.GATEWAY));
+        when(stack.getAllAvailableInstances()).thenReturn(instanceMetaDataAsList);
+        Optional<String> errorMessageSegment = NO_ERROR;
+        if (errorMessageSegment.isPresent()) {
+            checkExecutableThrowsExceptionAfterChecking(errorMessageSegment, () -> underTest.validateCMStatus(stack));
+        } else {
+            assertDoesNotThrow(() -> underTest.validateCMStatus(stack));
+        }
+    }
+
+    @Test
+    public void testValidateCMStatusAllGatewayHealthy() {
+        StackDto stack = mock(StackDto.class);
+        List<InstanceMetadataView> instanceMetaDataAsList = new ArrayList<>();
+        instanceMetaDataAsList.add(generateInstanceMetadata(InstanceStatus.SERVICES_HEALTHY, InstanceMetadataType.GATEWAY_PRIMARY));
+        instanceMetaDataAsList.add(generateInstanceMetadata(InstanceStatus.SERVICES_HEALTHY, InstanceMetadataType.GATEWAY));
+        when(stack.getAllAvailableInstances()).thenReturn(instanceMetaDataAsList);
+        Optional<String> errorMessageSegment = NO_ERROR;
+        if (errorMessageSegment.isPresent()) {
+            checkExecutableThrowsExceptionAfterChecking(errorMessageSegment, () -> underTest.validateCMStatus(stack));
+        } else {
+            assertDoesNotThrow(() -> underTest.validateCMStatus(stack));
+        }
+    }
+
+    @Test
+    public void testValidateCMStatusAllGatewayUnhealthy() {
+        StackDto stack = mock(StackDto.class);
+        List<InstanceMetadataView> instanceMetaDataAsList = new ArrayList<>();
+        instanceMetaDataAsList.add(generateInstanceMetadata(InstanceStatus.SERVICES_UNHEALTHY, InstanceMetadataType.GATEWAY_PRIMARY));
+        instanceMetaDataAsList.add(generateInstanceMetadata(InstanceStatus.SERVICES_UNHEALTHY, InstanceMetadataType.GATEWAY));
+        when(stack.getAllAvailableInstances()).thenReturn(instanceMetaDataAsList);
+        Optional<String> errorMessageSegment = Optional.of("Upscale is not allowed because the CM host is not healthy: hg0-host-2: SERVICES_UNHEALTHY.");
+        if (errorMessageSegment.isPresent()) {
+            checkExecutableThrowsExceptionAfterChecking(errorMessageSegment, () -> underTest.validateCMStatus(stack));
+        } else {
+            assertDoesNotThrow(() -> underTest.validateCMStatus(stack));
+        }
```

Comment:
For this test and the one below, I'm not sure why you're adding an if statement to make assertions for expected exceptions. You can directly pass the `errorMessageSegement` to `checkExecutableThrows..` instead of an `Optional`, and make assertions for scenarios where an exception is expected.

Reply:
Done